pkg/storage/util_test.go

```diff
@@ -143,10 +146,23 @@ func (m *mockChunkStore) Get(ctx context.Context, userID string, from, through m
 }
 
 func (m *mockChunkStore) GetChunkRefs(ctx context.Context, userID string, from, through model.Time, matchers ...*labels.Matcher) ([][]chunk.Chunk, []*chunk.Fetcher, error) {
-	refs := make([]chunk.Chunk, 0, len(m.chunks))
-	// transform real chunks into ref chunks.
-	for _, c := range m.chunks {
-		r, err := chunk.ParseExternalKey("fake", c.ExternalKey())
+	var chunks [][]chunk.Chunk
+	var fetcher []*chunk.Fetcher
+	chunksWithFetcher := []int{4, 1, 2}
+	var count int
+	for _, i := range chunksWithFetcher {
+		f, refs := m.buildFetchers(i, count)
+		fetcher = append(fetcher, f)
+		chunks = append(chunks, [][]chunk.Chunk{refs}...)
+		count = count + i
+	}
+	return chunks, fetcher, nil
+}
+
+func (m mockChunkStore) buildFetchers(chunksLength int, count int) (*chunk.Fetcher, []chunk.Chunk) {
+	refs := make([]chunk.Chunk, 0, chunksLength)
+	for i := count; i < chunksLength+count; i++ {
+		r, err := chunk.ParseExternalKey("fake", m.chunks[i].ExternalKey())
```

Comment:
I know this is not clean coding. However, this was the best and easiest I could think of.
Kindly guide me here if there are any other better ways

Reply:
I suggest you write a test for only this function `fetchLazyChunks(ctx context.Context, chunks []*chunkenc.LazyChunk)`

The only interface you need to implement then is a fetcher. You can have 2 implementation of fetcher one that returns an error and one that returns the chunk.
Then use that in the test and verify that for invalid checksum errors they are stamped as invalid.


Same goes for `buildHeapIterator(ctx context.Context, chks [][]*chunkenc.LazyChunk, filter logql.LineFilter, direction logproto.Direction, from, through time.Time) (iter.EntryIterator, error)` you could write a test only for this function and verify that lazychunk that are invalid are skipped.

Now your version is also good, it test both at the same time, if you just want to improve it. Then create a seperate mock for your own usage and separate test cases, feel free to copy or create a function for code that could be reuse between your new test function and the current one.

TLDR; the reason why it's ugly is because you're mixin different type of tests in the same table tests.

Reply:
Have added separate test for buildHeapIterator. 
I couldn't add for fetchLazyChunks since `FetchChunks` is not an interface.